### PR TITLE
TreeDDS: avoid unnecessary rebasing

### PR DESCRIPTION
**Pull request by yann-achard-MS**
This PR adds a check in the `EditManager` to avoid some unnecessary rebasing.
In the long run, this is just a performance improvement (albeit a possibly important one).
In the short run, this helps avoid cases where our current lack of support for true inverses causes unexpected merge outcomes.

### CODE REVIEW

File: packages/dds/tree/src/test/edit-manager/editManager.spec.ts [hunk 1]
```diff
@@ -375,6 +376,28 @@ describe("EditManager", () => {
...
+                changeset: TestChangeRebaser.mintChangeset([], 1),
+            };
+            const commit2: TestCommit = {
+                sessionId: peer1,
+                seqNumber: brand(2),
+                refNumber: brand(0),
+                changeset: TestChangeRebaser.mintChangeset([1], 2),
+            };
+            manager.addSequencedChange(commit1);
+            family.rebaser.invert = () => assert.fail("invert was called");
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/edit-manager/editManager.spec.ts:395]: Is it too inconvenient to create a separate ChangeFamily for this test instead of mutating the existing one?

File: packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts [hunk 2]
```diff
@@ -106,5 +106,45 @@ describe("SharedTree", () => {
                 tree2.forest.forgetAnchor(destination);
             }
         });
+
+        it("can insert multiple nodes", async () => {
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:110]: Is this test supposed to be part of this PR?

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts (hunk 2 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:110]: I was on the fence about it. It's how I ran into the issue (the test fails without this new code) so I figured it doesn't hurt to keep it.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:110]: Happy to take it out if you find it distasteful.

File: packages/dds/tree/src/test/edit-manager/editManager.spec.ts (hunk 1 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/test/edit-manager/editManager.spec.ts]: Good point. Turns out it wasn't too bad.

File: packages/dds/tree/src/edit-manager/editManager.ts [hunk 3]
```diff
@@ -241,6 +262,28 @@ export class EditManager<TChangeset, TChangeFamily extends ChangeFamily<any, TCh
 interface Branch<TChangeset> {
     localChanges: Commit<TChangeset>[];
     refSeq: SeqNumber;
+    /**
+     * True iff the commits on the branch were not based on the commit at trunk tip at the time
+     * they were sequenced.
```
Comment (alex-pardes) [inline packages/dds/tree/src/edit-manager/editManager.ts:267]: I find this phrasing a bit confusing. The meaning of "at the time they were sequenced" seems somewhat nebulous to me. While this could be reworded more clearly, I would also consider some alternative definitions. I think these are equivalent to your definition:

A branch is divergent iff it has local changes and there is a change with seqNumber between the branch's refSeq and the seqNumber of the last change in the branch.

In other words, the ref commit followed by the local changes forms a contiguous block in the trunk or final sequence.

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/edit-manager/editManager.ts (hunk 3 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/edit-manager/editManager.ts:267]: Regarding "and there is a change with seqNumber between the branch's refSeq and the seqNumber of the last change in the branch": isn't true of all the changes in the branch, not just the last one?

Regarding: "the ref commit followed by the local changes forms a contiguous block in the trunk": did you mean "does not form a contiguous block"?

Comment (alex-pardes) [reply packages/dds/tree/src/edit-manager/editManager.ts]: I meant to say:

A branch is divergent iff it has local changes and there is a change _on the trunk but not in the branch_ with seqNumber between the branch's refSeq and the seqNumber of the last change in the branch.

And yes, I dropped a negation at the end.

File: packages/dds/tree/src/edit-manager/editManager.ts [hunk 4]
```diff
@@ -241,6 +262,30 @@ export class EditManager<TChangeset, TChangeFamily extends ChangeFamily<any, TCh
 interface Branch<TChangeset> {
     localChanges: Commit<TChangeset>[];
     refSeq: SeqNumber;
+    /**
+     * A branch is divergent iff it has local changes and there is a change with a `seqNumber`
+     * between the branch's `refSeq` and the `seqNumber` of the last change in the branch.
```
Comment (alex-pardes) [inline packages/dds/tree/src/edit-manager/editManager.ts:267]:
```suggestion
     * A branch is divergent iff it has local changes and there is a change outside the branch with a `seqNumber`
     * between the branch's `refSeq` and the `seqNumber` of the last change in the branch.
```

File: packages/dds/tree/src/edit-manager/editManager.ts [hunk 5]
```diff
@@ -241,6 +262,30 @@ export class EditManager<TChangeset, TChangeFamily extends ChangeFamily<any, TCh
...
+     * In other words, the ref commit followed by the local changes
+     * do not from a contiguous block in the trunk or final sequence.
```
Comment (alex-pardes) [inline packages/dds/tree/src/edit-manager/editManager.ts:269]:
```suggestion
     * do not form a contiguous block in the trunk or final sequence.
```

**Review by alex-pardes — APPROVED** [review]
